File: DistanceSensing/UltraSound/sensor.py
import RPi.GPIO as GPIO
import serial
import time

class MB1040:
    # This function should only run once, when device starts
    def __init__(self):
        
        self.bit_delay = 104
        self.half_bit_delay =  52
        self.num_bytes = 5

        GPIO.setmode(GPIO.BOARD)
        self.rx = 10
        self.tx = 8

        GPIO.setup(self.rx, GPIO.IN)
        GPIO.setup(self.tx, GPIO.OUT)

        GPIO.output(self.tx, GPIO.HIGH)

        self.ser = serial.Serial('/dev/ttyAMA0')
        self.usleep = lambda x: time.sleep(x/1000000.0)
       

    def getDistance(self):
        data = []
        invalid_reading = self.readSonar(data)
        
        if not invalid_reading:
            self.printSonarMessage(data)

        self.usleep(100) 


    def end(self):
        self.ser.close()
        GPIO.cleanup()


    def readSonar(self, data):
        i = 0
        invalid = False

        data.append(self.readSonarByte())        

        if (data[0] == 'R'):
            for i in range(0, self.num_bytes):
                data.append(self.readSonarByte())
        else:
            invalid = True

        return invalid


    def readSonarByte(self):
        char = 0
        print(self.ser.read())

        while self.ser.read() == 0: 
            print("Waiting...")
        
        
        if self.ser.read() == GPIO.HIGH:
            self.usleep(self.half_bit_delay)

            for bit_shift in range(0, 7):
                self.usleep(self.bit_delay)
                char |= ((not self.ser.read()) << bit_shift)

            self.usleep(self.bit_delay)

        return char


    def printSonarMessage(self, msg):
        msg[self.num_bytes] = 0
        print(msg)


    def printSonarData(self, data):
        print("\tchar\thex\tdec\tbin")

        for i in range(0, self.num_bytes):
            print(i, "\t", data[i], "\t", end="") #char
            
            #hex
            data_hex = ''.join([hex(ord(x))[2:].zfill(2) for x in data[i]])
            print(data_hex, "\t", end="")
            
            print(ord(data[i]), "\t", end="") #decimal
           
            #binary
            data_bin = ''.join(format(ord(x), 'b') for x in data[i])
            print(data_bin)


mb = MB1040()
mb.getDistance()
#getDistance()

File: DistanceSensing/UltraSound/sensor.c
#include <stdlib.h>
#include <stdio.h>
#include <wiringSerial.h>
#include <wiringPi.h>

#define BIN_PATTERN "%c%c%c%c%c%c%c%c"
#define BYTE_TO_BINARY(byte) \
	(byte & 0x80 ? '1' : '0'), \
	(byte & 0x40 ? '1' : '0'), \
	(byte & 0x20 ? '1' : '0'), \
	(byte & 0x10 ? '1' : '0'), \
	(byte & 0x08 ? '1' : '0'), \
	(byte & 0x04 ? '1' : '0'), \
	(byte & 0x02 ? '1' : '0'), \
	(byte & 0x01 ? '1' : '0')


#define PORT "/dev/ttyAMA0"
#define BAUD 9600
#define HALF_BIT_DELAY 52
#define BIT_DELAY 104
#define NUM_BYTES 5
#define rx 10
#define tx 8

void loop();
bool readSonar(char []);
void readSonarByte(char *);
void printSonarMessage(char []);
void printSonarData(char []);


int fd;

int main() {
	fd = serialOpen(PORT, BAUD);
	wiringPiSetup();
	pinMode(rx, INPUT);
	pinMode(tx, OUTPUT);
	
	digitalWrite(tx, HIGH);

	loop();

	return 0;
} 

void loop() {
	char range[NUM_BYTES];
	bool invalidReading = readSonar(range);

	if (!invalidReading) 
		printSonarMessage(range);

	delay(100);
}

bool readSonar(char range[]) {
	int i = 0;
	bool invalid = false;
	
	readSonarByte(&range[0]);

	if (range[0] == 'R') {

		while (i < NUM_BYTES) 
			readSonarByte(&range[++i]);

	} else invalid = true;

	return invalid;
}

void readSonarByte(char *value) {
	*value = 0;

	while (digitalRead(rx) == 0);

	if (digitalRead(rx) == HIGH) {
		delayMicroseconds(HALF_BIT_DELAY);
		
		for (int bitShift = 0; bitShift < 7; bitShift++) {
			delayMicroseconds(BIT_DELAY);
			*value |= ((!digitalRead(rx)) << bitShift);
		}
		delayMicroseconds(BIT_DELAY);
	}
}

void printSonarMessage(char range[]) {
	range[NUM_BYTES-1] = 0;
	printf("%s\n", range);
}


void printSonarData(char range[]) {
	printf("\tCHAR\tHEX\tDEC\tBIN\n");
	for (int i = 0; i < NUM_BYTES; i++) {
		printf("%i\t", i);
		printf("%c\t", range[i]);
		printf("%02X\t", range[i]);
		printf("%d\t", (int) range[i]);
		printf(BIN_PATTERN, BYTE_TO_BINARY(range[i]));
		printf("\n");
	}
}






